### PR TITLE
Return slurm results

### LINKED ISSUE #197: Retrieve results of the SLURM simulation

### PR DESCRIPTION
**Pull request by p1003**

### CODE REVIEW

File: yaptide/routes/utils/response_templates.py [hunk 1]
```diff
@@ -2,20 +2,19 @@
...
 
-def yaptide_response(message: str, code: int, content: Union[dict, str] = "") -> Response:
+def yaptide_response(message: str, code: int, content: dict = {}) -> Response:
     """Function returning Response object"""
     response_dict = {'message': message}
-    if type(content) is dict:
-        for key, value in content.items():
-            response_dict[key] = value
+    for key, value in content.items():
+        response_dict[key] = value
```
Comment (grzanka) [inline yaptide/routes/utils/response_templates.py]:
```suggestion
    response_dict.update(content)
```

See https://docs.python.org/3.10/library/stdtypes.html#dict.update

**Comment by grzanka** [comment]
@p1003 when do you foresee this PR to be ready for review ?

**Comment by p1003** [comment]
@grzanka  I named PR incorrectly - now it has proper name

File: plgrid/string_templates.py [hunk 2]
```diff
@@ -40,3 +40,7 @@
 
 exit 0
 """  # skipcq: FLK-E501
+
+plgdata_list_url: str = """{{http_plgdata}}/list/{{hostname}}/net/people/{{plguserlogin}}/sh_output/{{job_id}}"""
```
Comment (grzanka) [inline plgrid/string_templates.py]: The path here needs to be adjustr to $SCRATCH
we shouldn't save simulations in $HOME

File: plgrid/string_templates.py [hunk 3]
```diff
@@ -40,3 +40,7 @@
 
 exit 0
 """  # skipcq: FLK-E501
+
+plgdata_list_url: str = """{http_plgdata}/list/{hostname}/net/people/{plguserlogin}/sh_output/{job_id}"""
```
Comment (retkiewi) [inline plgrid/string_templates.py]: Public consts should be uppercase.

File: yaptide/examples/call_api_example.py [hunk 4]
```diff
@@ -182,20 +183,26 @@ def run_simulation_with_files(session: requests.Session, example_dir, json_to_se
                 print(e)
 
 
-def run_simulation_with_rimrock(port: int = 5000):
-    """Example function running simulation on rimrock"""
-    example_dir = os.path.dirname(os.path.realpath(__file__))
-    grid_proxy_path = Path(example_dir, 'grid_proxy')
+def get_grid_proxy_file(dir_path: str) -> str:
```
Comment (retkiewi) [inline yaptide/examples/call_api_example.py]:
```suggestion
def get_grid_proxy(dir_path: str) -> str:
```
or
```suggestion
def read_grid_proxy_file(dir_path: str) -> str:
```
seems like a more accurate name

File: yaptide/examples/call_api_example.py [hunk 5]
```diff
@@ -182,20 +183,26 @@ def run_simulation_with_files(session: requests.Session, example_dir, json_to_se
...
+    """Function reading grid_proxy file"""
+    grid_proxy_path = Path(dir_path, 'grid_proxy')
```
Comment (retkiewi) [inline yaptide/examples/call_api_example.py:188]: Is this variable needed?

File: yaptide/examples/call_api_example.py [hunk 6]
```diff
@@ -182,20 +183,26 @@ def run_simulation_with_files(session: requests.Session, example_dir, json_to_se
...
         cmd += f" > {grid_proxy_path} && unset p\n"
         print(cmd)
-        return
+        exit(1)
+
+
+def run_simulation_with_rimrock(port: int = 5000):
+    """Example function running simulation on rimrock"""
+    example_dir = os.path.dirname(os.path.realpath(__file__))
+    grid_proxy = get_grid_proxy_file(dir_path=example_dir)
```
Comment (retkiewi) [inline yaptide/examples/call_api_example.py]: No need to specify the function parameter name as it has only one named parameter.
```suggestion
    grid_proxy = get_grid_proxy_file(example_dir)
```

File: yaptide/examples/call_api_example.py [hunk 7]
```diff
@@ -208,25 +215,57 @@ def run_simulation_with_rimrock(port: int = 5000):
     job_id: str = ""
     job_id = res_json.get('job_id')
```
Comment (retkiewi) [inline yaptide/examples/call_api_example.py:219]:
```suggestion
    job_id: str = res_json.get('job_id')
```

File: yaptide/examples/call_api_example.py [hunk 8]
```diff
@@ -208,25 +215,57 @@ def run_simulation_with_rimrock(port: int = 5000):
...
+
+
+def get_slurm_results(port: int = 5000):
+    """Example function getting slurm results"""
+    example_dir = os.path.dirname(os.path.realpath(__file__))
+    grid_proxy = get_grid_proxy_file(dir_path=example_dir)
+
+    headers = {"PROXY": base64.b64encode(grid_proxy.encode('utf-8')).decode('utf-8')}
+    session = requests.Session()
+    job_id = "854704.ares.cyfronet.pl"
```
Comment (retkiewi) [inline yaptide/examples/call_api_example.py]: Why is the job_id hardcoded? Won't it change?

File: yaptide/examples/call_api_example.py [hunk 9]
```diff
@@ -208,25 +215,57 @@ def run_simulation_with_rimrock(port: int = 5000):
...
+    plguserlogin = "plgpitrus"
```
Comment (retkiewi) [inline yaptide/examples/call_api_example.py]: Should be a repository secret

File: yaptide/examples/xd.txt [hunk 10]
```diff
@@ -0,0 +1,9 @@
+curl -X GET https://data.plgrid.pl/list/ares/net/people/plgpitrus/ -H "PROXY:`cat grid_proxy | base64 | tr -d '\n'`"
```
Comment (retkiewi) [inline yaptide/examples/xd.txt]: Nice filename ;)

File: plgrid/plgdata_methods.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+from pathlib import Path
+
+from plgrid.string_templates import plgdata_get_url, plgdata_list_url
+
+from pymchelper.input_output import fromfile
+
+from yaptide.celery.tasks import pymchelper_output_to_json
+
+http_plgdata = 'https://data.plgrid.pl'
+hostname = 'ares'
```
Comment (retkiewi) [inline plgrid/plgdata_methods.py:14]: Probably should not be hardcoded in the script

Comment (p1003) [reply plgrid/plgdata_methods.py:14]: PR is still draft -> this will be changed

File: yaptide/examples/call_api_example.py (hunk 9 above)
Comment (p1003) [reply yaptide/examples/call_api_example.py]: Yes, it should be

File: plgrid/plgdata_methods.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def fetch_bdo_files(json_data: dict) -> tuple[dict, int]:
+    """Function fetching result bdo files from cluster"""
```
Comment (grzanka) [inline plgrid/plgdata_methods.py]:
```suggestion
    """
    Fetching simulation results from the cluster.
    Return a tuple with a dictionary containing results and an integer error code.
    """
```

File: plgrid/plgdata_methods.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    session = requests.Session()
+    headers = {
+        "PROXY": json_data['grid_proxy']
+    }
+    job_id: str = json_data['job_id'].split('.')[0]
```
Comment (grzanka) [inline plgrid/plgdata_methods.py]: Why do you need to split on `.`. Provide a comment with example format of this

File: plgrid/plgdata_methods.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    list_url = plgdata_list_url.format(
```
Comment (grzanka) [inline plgrid/plgdata_methods.py]:
```suggestion
    // construct an URL with a path to the simulation workspace directory, based on jobid
    // this path may contain many different files: simulation output (i.e. BDO files), logs and simulation output
    list_url = plgdata_list_url.format(
```

File: plgrid/plgdata_methods.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        http_plgdata=http_plgdata,
+        hostname=hostname,
+        job_id=job_id,
+    )
+    res: requests.Response = session.get(list_url, headers=headers)
```
Comment (grzanka) [inline plgrid/plgdata_methods.py:36]:
```suggestion
    // get a list of all files in the simulation workspace directory
    res: requests.Response = session.get(list_url, headers=headers)
```

File: plgrid/plgdata_methods.py (hunk 14 above)
Comment (grzanka) [reply plgrid/plgdata_methods.py]: @p1003 is this comment true ?

File: plgrid/string_templates.py [hunk 16]
```diff
@@ -40,3 +40,7 @@
 
 exit 0
 """  # skipcq: FLK-E501
+
+plgdata_list_url: str = """{http_plgdata}/list/{hostname}/~/sh_output/{job_id}"""
```
Comment (grzanka) [inline plgrid/string_templates.py]:
```suggestion
// as for now we hardcode that the results are being saved in home directory
plgdata_list_url: str = """{http_plgdata}/list/{hostname}/~/sh_output/{job_id}"""
```

File: plgrid/plgdata_methods.py (hunk 14 above)
Comment (p1003) [reply plgrid/plgdata_methods.py]: it also contains dictionaries

File: plgrid/plgdata_methods.py (hunk 13 above)
Comment (p1003) [reply plgrid/plgdata_methods.py]: added comment

File: plgrid/string_templates.py (hunk 2 above)
Comment (p1003) [reply plgrid/string_templates.py]: Let's work on it a bit later

**Review by grzanka — APPROVED** [review]
LGTM

**Pull request merged**